src/macros-by-example.adoc

```diff
@@ -0,0 +1,827 @@
+single stream, and the expander verifies that each value on that stream is acceptable by the
+parameter’s declared type.
+
+----
+(:prices (1 (:values 2 3) 4) GBP) ⇒ {amount:1, currency:GBP}
+                                    {amount:2, currency:GBP}
+                                    {amount:3, currency:GBP}
+                                    {amount:4, currency:GBP}
+----
+
+
+==== one-or-more
+
+A one-or-more `*+*` parameter works just like `***` except the resulting stream must produce at
+least one value.  To continue abusing our prices example:
+
+[{mrk}]
+----
+(*public* prices
+  [(*number+* amount), (*symbol* currency)]
+  (*for_each* amount
+    (price amount currency)))
+----
+
+[{mrk}]
+----
+(:prices () JPY) ⇒ _**error**: at least one value expected for + parameter_
+(:prices [99] EUR)               ⇒ {amount:99, currency:EUR}
+(:prices \((:) (:values 99)) EUR) ⇒ {amount:99, currency:EUR}
+----
+
+Note that among the delimited expressions, empty results are fine, as long as their concatenation
+produces one or more values.
+
+
+=== Optional Arguments
+
+When a trailing parameter is voidable, an invocation can omit its corresponding argument, as long
+as no following parameter is being given a value.  We’ve seen this as applied to rest-parameters,
+but it also applies to `*?*` and `***` cardinality:
+
+[{mrk}]
+----
+(*public* optionals
+  [(**any* **a), (*any?* b), (*any!* c), (*any** d), (*any?* e), (*any\...* f)]
+  (make_list a b c d e f))
+----
+
+Since `d`, `e`, and `f` are all voidable, they can be omitted by invokers.  But `c` is required so
+`a` and `b` must always be present, at least as an empty delimiter:
+
+----
+(:optionals [] (:) for_c) ⇒ [for_c]
+----
+
+Now `c` receives the symbol `for_c` while the other parameters are all void.  If we want to provide
+just `e`, then we must also provide a delimiter for `d`:
+
+----
+(:optionals [] (:) for_c () for_e) ⇒ [for_c, for_e]
+----
+
+
+=== Tagless and Fixed-Width Types
+
+It’s time to shift gears to discuss a feature that primarily benefits binary-encoded data:
+tagless types.
+
+Those familiar with binary Ion will understand that the encoding of every value starts off with a
+“type tag”, an opcode that indicates the data-type of the next value and thus the interpretation
+of the following octets of data.  In general, these tags also indicate whether the value has
+annotations, and whether it’s null.
+
+These tags are necessary because the Ion data model allows values of any type to be used
+anywhere. Ion documents are not schema-constrained: nothing forces any part of the data to have a
+specific type or shape.  We call Ion “self-describing” precisely because each value
+self-describes its type via a type tag.
+
+This adds considerable overhead throughout the document, which is frustrating when schema
+constraints are enforced through some mechanism outside the serializer/deserializer.  It’s worse
```

Comment:
Editorial suggestion—we should focus on technical reasons rather than feeling (unless we're specifically talking about developer experience/ease-of-use).
```suggestion
If schema constraints are enforced through some mechanism outside the serializer/deserializer, the type tags are unnecessary and may add up to a non-trivial amount of wasted space.
```